File: fixtures/projects/e2e-python/src/game/session.py
class Session:
    def __init__(self, match_id: str, players: list[str]) -> None:
        self.match_id = match_id
        self.players = players
        self.round_scores: list[int] = []
        self.active = False

    def start_match(self) -> dict[str, int | str]:
        if len(self.players) < 2:
            raise ValueError("a game session needs at least two players")
        self.active = True
        self.round_scores.clear()
        return {"match_id": self.match_id, "player_count": len(self.players)}

    def end_round(self, score: int) -> dict[str, int]:
        if not self.active:
            raise RuntimeError("cannot end a round before the match starts")
        self.round_scores.append(score)
        return {"round": len(self.round_scores), "total_score": sum(self.round_scores)}

    def finish_match(self) -> dict[str, int | str]:
        winner_index = len(self.round_scores) % len(self.players)
        self.active = False
        return {
            "winner": self.players[winner_index],
            "rounds_played": len(self.round_scores),
        }


def open_match(match_id: str, players: list[str]) -> Session:
    session = Session(match_id=match_id, players=players)
    session.start_match()
    return session

File: fixtures/projects/e2e-app/src/game/session.ts
export class Session {
	private rounds: number[] = [];
	private active = false;

	constructor(public readonly matchId: string, public readonly players: string[]) {}

	start(): { matchId: string; players: number } {
		if (this.players.length < 2) {
			throw new Error("A game session needs at least two players");
		}
		this.active = true;
		this.rounds = [];
		return { matchId: this.matchId, players: this.players.length };
	}

	endRound(score: number): { round: number; total: number } {
		if (!this.active) {
			throw new Error("Cannot end a round before the match session starts");
		}
		this.rounds.push(score);
		return {
			round: this.rounds.length,
			total: this.rounds.reduce((sum, value) => sum + value, 0),
		};
	}

	finish(): { winner: string; roundsPlayed: number } {
		const bestIndex = this.rounds.length % this.players.length;
		this.active = false;
		return { winner: this.players[bestIndex] ?? this.players[0], roundsPlayed: this.rounds.length };
	}
}

export function startGame(matchId: string, players: string[]): Session {
	const session = new Session(matchId, players);
	session.start();
	return session;
}

export function endRound(session: Session, score: number): { round: number; total: number } {
	return session.endRound(score);
}